### PR TITLE
PCHR-2454: Changes to documents filtering on documents dashboard

### PR DESCRIPTION
**Pull request by bmnepali**
## Overview
This PR add the fixes requested as in this Pdf: 
[filters (3).pdf](https://github.com/compucorp/civihr-tasks-assignments/files/1382338/filters.3.pdf)

1. The status filter field is moved form sidebar to advanced filter section in /documents
2. The Left sidebar now has "All" filter which will filter documents regardless of due or expiry date of the document.
3. The sidebar date range section will change it's background if the any of the date range is selected.
4. The status text of the selected status in advanced filters is change to "Capital Case" format/case.

## Before
The following gif shows following points:
- The status filter was in sidebar.
- Left sidebar did not have "All" filter
- The selected status text was "UPPERCASE"
- Date range filter section didn't change the background on selecting dates
2454-before-iloveimg-compressed

### Reopened for following contact field overlapping issue:
screen shot 2017-10-16 at 14 59 20

## After
The following gif shows following points:
- The status filter was in now in advanced filter section.
- Left sidebar now has "All" filter nav tab
- The selected status text was "Capital Case"
- Date range filter section changes its background on selecting dates
2454-final-iloveimg-compressed

The final solution after QA:
- The alignment of dates filter section is improved
- The background of the select dates filter section is changed
- The auto resizing of the select option for contact in advanced filter is removed.
2454-final-iloveimg-compressed

#### Fixed the contact field overlapping assignment field issue by limiting the characters in the contact field.
2454-finalee

## Technical Details
 1. Update/ add new filters in left sidebar as below:
> Overdue
> Next fortnight
> 90 days
> All

2. Remove status filter form sidebar to the filters in document list section.
```html
<div class="{{prefix}}filter-advanced" uib-collapse="list.isCollapsed.filterAdvanced">
  <div class="row form-horizontal">
    <div class="col-xs-4">
      ... // contacts filters
    </div>
    <div class="col-xs-4">
      ... // Assignment filter
    </div>
    <div class="col-xs-4">
      <div class="form-group">
        <label class="col-xs-12 col-sm-4 control-label">Assignment Type:</label>
        <div class="col-xs-12 col-sm-8">
          <ui-select  prevent-animations multiple no-dropdown ng-model="list.filterParams.assignmentType">
            <ui-select-match prevent-animations class="ui-select-match" placeholder="Select assignment...">{{$item.title}}</ui-select-match>
            <ui-select-choices prevent-animations class="ui-select-choices" repeat="type.id as type in cache.assignmentType.arr">
              <div ng-bind-html="type.title | highlight: $select.search"></div>
            </ui-select-choices>
          </ui-select>
        </div>
      ...
 </div>
```

3. Update the filters list in document list controller:
Filter params:
```js
vm.dueFilters = [
      ...
      { badgeClass: 'primary', calendarView: 'month', value: 'all' }
];
```
Filter labels:
```js
vm.label = {
     ...
     dueInNextFortnight: 'Next fortnight',
     dueInNinetyDays: ' 90 days',
     all: 'All',
     ...
}
```

4. Display the value filters based on wether the the filter "All" is selected or not. If "All"nfilter is selected then list all the documents without filtering. 
```html
<ul class="nav nav-pills nav-stacked">
    ...
        {{list.label[filter.value]}}
        <i class="fa fa-calendar" ng-click="list.viewInCalendar(filter.calendarView)"></i>
        <span class="badge badge-{{filter.badgeClass}}" ng-if="filter.value !== 'all'">
          {{(
            (list.filterByDateField(filter.value))
            | filterByStatus:list.filterParams.documentStatus || ''
            | filterByOwnership:list.filterParams.ownership
            | filterByAssignmentType:list.filterParams.assignmentType
            ).length
          }}
        </span>
        <span class="badge badge-{{filter.badgeClass}}" ng-if="filter.value === 'all'">
          {{(list.list).length}}
        </span>
      </a>
    </li>
  </ul>
```

5. The width of Advanced filters section is limited to "817px;", remove it so the width spans to 100% on the parent. in file "/scss/civitasks/partials/_filters.scss"
```css
.#{$prefix}filter-advanced {
  // max-width: 817px; <- Remove 
  ...
}
```

6. To change the background of the date range filter. Wrap the date range fields in a div and change its background when the fields are selected:
```html
<div ng-class="{'nav date-range__selected': (list.filterParamsHolder.dateRange.from || list.filterParamsHolder.dateRange.until)}">
    ... // Date range fields
</div>
```
**Note:** But since the sidebar filter is shared between tasks and documents. This means the style(background) will change if date range section will change when dates are selected in tasks as well which is not expected. So, we need to differentiate which view is displayed currently. For this add a property in document list controller as:
```js
vm.isDocumentSection = false;
```

changing the value of `isDocumentSection` when the filter is applied:
``js
function applySidebarFilters () {
    ...
    vm.isDocumentSection = (vm.filterParamsHolder.dateRange.from !== null || vm.filterParamsHolder.dateRange.until !== null);
    ...
}
```
So the code now becomes:
```html
<div ng-class="{'nav date-range__selected': list.isDocumentSection}">
   ...
</div>
```
That property will help us decide wether to add the newly created class "date-range__selected" to be added in task filters or not.

7. And define the styles for the background creating '.date-range__selected' css class in "/scss/civitasks/partials/_sidebar.scss" file.
```css
.#{$prefix}sidebar-inner {
  ...
  .date-range__selected {
    .#{$prefix}select-dates{
     ...
    }

    .form-horizontal {
     ...
    }
  }
}
```
8. Changes the status in dcument-list.controller.js to set the date filters to collapsed by default:
```js
vm.isCollapsed = {
     ...
     filterDates: true
};
```
9. Make the text of select options dropdown to "Capital Case"font case. In file "/scss/civitasks/partials/_filters.scss"
```css
  .ui-select-match-item.btn {
    text-transform: capitalize;
  }
```
10. Remove `/src/tasks-assignments/controllers/document-list.js` in favor of using `/src/tasks-assignments/controllers/document-list.controller.js` file.

#### After reopened for the contact field overlapping the assignment field on large contact name.
screen shot 2017-10-16 at 14 59 20

11. Limit the characters to 32 in contact select option field to prevent resizing of the field in advanced filters in documents.
```html
<ui-select-match prevent-animations class="ui-select-match" placeholder="Search for staff member or assignee">
    {{ ($select.selected.label | limitTo: 32) + ($select.selected.label.length > 32 ? '...' : '') }}
</ui-select-match>
```

#### Few issues:
- The background of selected "OR Select dates" section is not correct (should be light blue)
- There is padding issues between the selected and not selected "OR Select dates" section.
- The background should only change when the filter is applied.
- The contact ui-select was resizing to the selected value length, which should not happen.

#### How its fixed:
- Changed the background color if the selected section
- Fixed the padding issue by amending the paddings.
- Made the `isDocumentSection` depend on date ranges and only change when the filter is applied
```js
function applySidebarFilters () {
    ...
    vm.isDocumentSection = (vm.filterParamsHolder.dateRange.from !== null || vm.filterParamsHolder.dateRange.until !== null);
    ...
}
```
- Refactored html markup, so the contact select does not get resized.

## Comments
- For QA:
When testing this fix, following pages are also to  be tested to mase sure that it didnot break any other things:
1. Tasks in T&A works as expected.
2. The changes introduced in '/documents' does not have any effect on '/tasks'
3. Filtering of documents still works
4. Filtering of tasks still works

## Tests
✅ Jasmine Tests - passed
✅ JS distributive files - created
✅ CSS distributive files - created
⏹ Backstop Tests - n/a
⏹ PHP Unit Tests - not needed to be run

### CODE REVIEW

File: uk.co.compucorp.civicrm.tasksassignments/views/directives/sidebar-filters.html [hunk 1]
```diff
@@ -1,10 +1,10 @@
 <div class="{{prefix}}filter-date">
   <ul class="nav nav-pills nav-stacked">
     <li role="presentation" ng-repeat="filter in list.dueFilters">
-      <a href ng-model="list.filterParams.due" uib-btn-radio="filter.value" uncheckable>
+      <a href ng-model="list.filterParams.due" uib-btn-radio="(filter.value == 'all') ? '' : filter.value" uncheckable>
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/views/directives/sidebar-filters.html]: Use `===` and `!==`

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/document-list.controller.js [hunk 2]
```diff
@@ -25,6 +25,7 @@ define([
     var vm = this;
     var defaultDocumentStatus = ['1', '2']; // 1: 'awaiting upload' | 2: 'awaiting approval
 
+    vm.calledFormDocuments = true;
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/document-list.controller.js]: A better name like `isDocumentSection` would be more readable. Also please add comment in the code itself, explanation what it does.

File: uk.co.compucorp.civicrm.tasksassignments/scss/civitasks/partials/_sidebar.scss [hunk 3]
```diff
@@ -1,3 +1,6 @@
+// Variables
+$light-gray : #D2D2D2;
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/scss/civitasks/partials/_sidebar.scss]: The variables should be moved to `sites/all/modules/civicrm/tools/extensions/civihr_tasks/uk.co.compucorp.civicrm.tasksassignments/scss/civihr/modules/_variables.scss` file.

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/document-list.controller.js (hunk 2 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/document-list.controller.js]: Agreed! will fix.

File: uk.co.compucorp.civicrm.tasksassignments/scss/civitasks/modules/_variables.scss [hunk 4]
```diff
@@ -1,7 +1,8 @@
 // Variables for Sidebar
-$light-gray : #D2D2D2;
-$secondary-colour: #707C88;
-$sidebar-main-bg: #fff;
+$light-blue:        #41AFCB;
+$white:             #FFFFFF;
+$secondary-colour:  #707C88;
+$sidebar-main-bg:   #fff;
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/scss/civitasks/modules/_variables.scss]: This is not necessary, as you already have $white.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js [hunk 5]
```diff
@@ -321,6 +321,28 @@ define([
       });
     });
 
+    describe('applySidebarFilters()', function () {
+      beforeEach(function () {
+        initController();
+      });
+
+      it('does not highlight the select dates filter section', function () {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: Organization could be better
`describe('before applySidebarFilters()' is called` and
`describe('after applySidebarFilters()' is called`

File: uk.co.compucorp.civicrm.tasksassignments/scss/civitasks/modules/_variables.scss (hunk 4 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/scss/civitasks/modules/_variables.scss]: Thanks for catching! fixed.

**Pull request merged**